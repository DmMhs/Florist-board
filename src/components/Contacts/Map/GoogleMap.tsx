import React, { Component, RefObject } from 'react';

import './GoogleMap.less';
import Spinner from '../../Spinner/Spinner';

interface GoogleMapProps {
  url: string;
}
interface GoogleMapState {
  mobileMode: boolean;
  mapIsFetching: boolean;
}

let resizeListener: EventListener;

class GoogleMap extends Component<GoogleMapProps, GoogleMapState> {
  private mapRef: RefObject<HTMLIFrameElement>;
  constructor(props: GoogleMapProps) {
    super(props);
    this.state = {
      mobileMode: false,
      mapIsFetching: true
    };
    this.mapRef = React.createRef();
  }
  public componentDidMount() {
    if (window.innerWidth < 768) {
      this.setState({
        mobileMode: true
      });
    } else {
      this.setState({
        mobileMode: false
      });
    }
    resizeListener = () => {
      if (window.innerWidth < 768) {
        this.setState({
          mobileMode: true
        });
      } else {
        this.setState({
          mobileMode: false
        });
      }
    };
    window.addEventListener('resize', resizeListener);
  }

  public componentWillUnmount() {
    window.removeEventListener('resize', resizeListener);
  }

  private mapFetchedHandler = () => {
    this.setState({
      mapIsFetching: false
    });
  };

  public render() {
    const { url } = this.props;
    const googleMap =
      this.state.mobileMode === true ? (
        <div
          className="mapouter Map"
          style={{
            position: 'relative',
            textAlign: 'right',
            height: '400px',
            width: '300px'
          }}
        >
          <div
            className="gmap_canvas"
            style={{
              overflow: 'hidden',
              background: 'none!important',
              height: '400px',
              width: '300px'
            }}
          >
            <iframe
              width="300"
              height="400"
              id="gmap_canvas"
              src={url}
              frameBorder="0"
              scrolling="no"
              marginHeight={0}
              marginWidth={0}
              onLoad={this.mapFetchedHandler}
            />
          </div>
        </div>
      ) : (
        <div
          className="mapouter Map"
          style={{
            position: 'relative',
            textAlign: 'right',
            height: '400px',
            width: '700px'
          }}
        >
          <div
            className="gmap_canvas"
            style={{
              overflow: 'hidden',
              background: 'none!important',
              height: '400px',
              width: '700px'
            }}
          >
            <iframe
              width="700"
              height="400"
              id="gmap_canvas"
              src={url}
              frameBorder="0"
              scrolling="no"
              marginHeight={0}
              marginWidth={0}
              onLoad={this.mapFetchedHandler}
              ref={this.mapRef}
            />
          </div>
        </div>
      );

    return (
      <div>
        {this.state.mapIsFetching === true ? <Spinner /> : null}
        {googleMap}
      </div>
    );
  }
}

export default GoogleMap;
