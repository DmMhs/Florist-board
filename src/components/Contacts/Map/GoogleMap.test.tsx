import React from 'react';
import { shallow } from 'enzyme';

import GoogleMap from './GoogleMap';
describe('GoogleMap works as expected', () => {
  it('GoogleMap component matches a snapshot', () => {
    const wrapper = shallow(<GoogleMap url="dumb-url" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('window size has impact on state', () => {
    global.innerWidth = 600;
    let wrapper = shallow(<GoogleMap url="dumb-url" />);
    let instance = wrapper.instance();
    expect(instance.state.mobileMode).toBeTruthy();
    global.innerWidth = 800;
    wrapper = shallow(<GoogleMap url="dumb-url" />);
    instance = wrapper.instance();
    expect(instance.state.mobileMode).toBeFalsy();
  });

  it('resize event listener changes the state', () => {
    const wrapper = shallow(<GoogleMap url="dumb-url" />);
    const instance = wrapper.instance();
    instance.setState({
      mobileMode: true
    });
    global.innerWidth = 800;
    window.dispatchEvent(new Event('resize'));
    expect(instance.state.mobileMode).toBeFalsy();
    global.innerWidth = 600;
    window.dispatchEvent(new Event('resize'));
    expect(instance.state.mobileMode).toBeTruthy();
  });

  it('mapFetchedHandler changes the state', () => {
    const wrapper = shallow(<GoogleMap url="dumb-url" />);
    const instance = wrapper.instance();
    instance.setState({
      mobileMode: true,
      mapIsFetching: true
    });
    instance.mapFetchedHandler();
    expect(instance.state.mapIsFetching).toBeFalsy();
  });
});
