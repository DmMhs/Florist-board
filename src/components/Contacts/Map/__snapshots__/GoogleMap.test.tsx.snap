// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoogleMap works as expected GoogleMap component matches a snapshot 1`] = `
<div>
  <WithStyles(CircularIndeterminate) />
  <div
    className="mapouter Map"
    style={
      Object {
        "height": "450px",
        "position": "relative",
        "textAlign": "right",
        "width": "724px",
      }
    }
  >
    <div
      className="gmap_canvas"
      style={
        Object {
          "background": "none!important",
          "height": "450px",
          "overflow": "hidden",
          "width": "724px",
        }
      }
    >
      <iframe
        frameBorder="0"
        height="450"
        id="gmap_canvas"
        marginHeight={0}
        marginWidth={0}
        onLoad={[Function]}
        scrolling="no"
        src="dumb-url"
        width="724px"
      />
    </div>
  </div>
</div>
`;
